breaking: make urql atoms behave as close to urql-react bindings as possible

RIP21 opened the pull request:
Changelog:
- Changed APIs to be as close as possible to the `urql` official bindings. 
- Mutations now are bound to only 1 query and will not suspend while fetching. Same way `useMutation` works in official bindings. Otherwise, it's quite unusable IMO. (covered by unit and playwright tests)
- Queries/Subscriptions no matter status or data (or `suspend: true` in client, will always suspend (I consider removing the data atom altogether as harmful API) (queries covered by tests, subscriptions are identical, so they aren't)
- Introduced Playwright tests as a more reliable and realistic testing technique. Especially in the case of suspense
- Unit tests that I kept are now using `vitest`. To avoid Typescript type-checking standing in a way of iterating on tests. The type-checking is still in place as a separate step. This also makes it so we don't type-check twice.
- Added `testing-app` that is used for playwright testing. It's not working without request mocking by the `playwright` on its own.


## Why?
Current bindings are extremely far from being correct. I don't know how widely it's used and if used in production. There is no way to reexecute queries (there is no such way as refetching in URQL, only reexecute of operation with different context), the current `refresh`only refreshes if the original context has `requestPolicy` being `network-only` or `cache-and-network` which is not how it suppose to be. It is also quite laggy causing hundreds of requests for some weird reason.
These changes fix it all. Since they're using all the `@urql/core` APIs as correctly as I managed to understand (by looking through multiple exchanges, original react bindings source code, etc.), it works perfectly in tandem with original react bindings of URQL or independently. Cache and everything get reused, refreshed, etc. no matter what binding causes changes Jotai or URQL.

Two things that I didn't managed to align with URQL official binding.
- in case of reexecute with `networkPolicy: "network-only" `operation result (status) atom is not reporting `{ stale: true }`. It just jumps directly to a new result. My debugging of `source` with `subscribe` and logging (or observable after conversion with `toObserver`), showed, that source actually emits it, but it's not coming to the component that uses that atom. Using `requestPolicy: "cache-and-network"` for refetch causes the reporting of `{ stale: true }` while refetching. In URQL official binding behavior is exactly the same for both `requestPolicies` while reexecuting.
- In case of error and redirect to an error boundary. If refetch (retry) is triggered from the error boundary. URQL official bindings will suspend the tree while fetching showing loader fallback. In the case of Jotai binding it immediately shows the errored component causing it to error again due to `if(opResult.error) throw opResult.error` in it, making it returning back to `ErrorBoundary` instantly in the loop unless guarded with `if(opResult.error && opResult.stale) return null` which is suboptimal, as Suspense fallback is preferable.

I need help from @dai-shi to figure out how to suspend it during that recovery refetch (@kitten may hint smth) and, also, possibly, from both of you, to figure out why stale status is not getting reported even though the source reports that to the subscriber.  It's important to notice, that `suspense:true/false` on urql client doesn't change either of these behaviors.
Also, @kitten, will be glad if you will review refetching and other logic related to urql, as I feel that some of it is a bit redundant.

@dai-shi If merged I'll gladly update docs with a migration guide and examples.

Closes: #10

RIP21 commented on tests/playwright-tests/smoke.test.ts:
```diff
@@ -0,0 +1,130 @@
+  // Reexecute with network-only policy, hence update
+  await page.getByText('refetch-network-only').click()
+  // Updated
+  // TODO: This is another problem. For some reason stale state on network-only
```
This is the case I'm talking about of refetching stale status not being reported.

RIP21 commented on tests/playwright-tests/smoke.test.ts:
```diff
@@ -0,0 +1,130 @@
+  // Since we redirect instantly on retry click, to avoid immediate render (and another throw of error that will get us back to error boundary)
+  // we have a stale-with-error short circuit, this ideally should be fixed to be suspended. As original urql client react bindings suspend the tree before
+  // results are in and then renders.
+  // TODO: Make it so it behaves similarly to URQL react binding
```
And this is the short circuit I'm talking about to avoid circular error throw as Jotai is not suspended the tree on retry from ErrorBoundary. Which is correct (as there is no some promise or so, to suspend the tree, but I'm dumb to figure out how to do that in jotai)

RIP21 commented on tests/test-app/src/SmokeTest.tsx:
```diff
@@ -0,0 +1,72 @@
+  const [{ error, stale, data }, dispatch] = useAtom(countStatus)
+
+  // This is needed to avoid return to the error boundary on retry. URQL react bindings suspend in this case. This needs to be fixed if possible.
+  if (stale && error) {
```
Here is the hack to be able to recover from the error boundary

RIP21 commented on tests/unit-tests/atomsWithMutation.spec.tsx:
```diff
@@ -57,38 +70,45 @@ it('mutation basic test', async () => {
 })
 
 describe('error handling', () => {
+  beforeAll(() => {
```
This is to remove spam into the console from error logs in these tests.

RIP21 commented on .eslintrc.json:
```diff
@@ -93,31 +94,12 @@
-        "map": [
-          ["^jotai-urql$", "./src/index.ts"]
-        ]
+      "eslint-import-resolver-custom-alias": {
```
There were issues resolving some modules (`vitest` to be precise) so this had to be used as older resolver is 4 years without any updates. 
As per: https://github.com/vitest-dev/vitest/issues/1827

dai-shi commented:
Wow, your work looks great.
We have been looking for a maintainer for `jotai-urql` and you should be the one.

dai-shi commented on src/clientAtom.ts:
```diff
@@ -1,5 +1,5 @@
 import { cacheExchange, createClient, fetchExchange } from '@urql/core'
-import { atom } from 'jotai/vanilla'
+import { atom } from 'jotai'
```
We should use `jotai/vanilla` which is the same thing, and allows to use `jotai-urql` without React.
```suggestion
import { atom } from 'jotai/vanilla'
```

dai-shi commented on src/common.ts:
```diff
@@ -14,18 +14,10 @@ export const createAtoms = <
-    client: Client,
-    refresh: () => void
-  ) => ActionResult
+  handleAction: (action: Action | Partial<OperationContext>) => ActionResult
```
I'm not sure if the union type is ideal, but it's your call.

dai-shi commented on src/common.ts:
```diff
@@ -14,18 +14,10 @@ export const createAtoms = <
 ) => {
   const refreshAtom = atom(0)
 
-  if (process.env.NODE_ENV !== 'production') {
-    refreshAtom.debugPrivate = true
-  }
-
```
Can't we keep it?

dai-shi commented on src/common.ts:
```diff
@@ -61,52 +50,26 @@ export const createAtoms = <
     }
   )
 
-  const baseDataAtom = atom((get) => {
```
baseDataAtom and dataAtom is the way to support suspense.
Can we keep the original structure?

RIP21 replied:
Unsure if it changes anything really as both status and data atoms are suspended currently.
Which is clearly seen in all tests even without this. Plus, unsure if old logic with filtering of all the signals depending on if there is error or not, is healthy. As I said in PR description, IMO I would get rid of data atom altogether as rather useless API.

RIP21 replied:
I'll test it out without data atom, so it will be only atomWithQuery/Subscription/Mutation. Will see if it changes anything. But now it gets suspended on the first load.
Weirdly enough it works even without a Suspense wrapper, just shows blank space before rendering entire thing, and I don't have any protection like save data access via optional chaining or guarding of any sort in the component, so from my understanding it should fail with "something is undefined" TypeError or similar, but it just suspends, but without any loading fallback (as if loading fallback is `() => null`, component)

dai-shi replied:
> Weirdly enough it works even without a Suspense wrapper

Maybe you use `atomWithObservable` without `initialData` for `statusAtom`.
It will trigger suspense for the initial load.

dai-shi reviewed (commented):
Reviewed clientAtom.ts and common.ts.

RIP21 commented:
So @dai-shi yesterday I added an API that uses `suspenseAtom` that will allow users to rehydrate atoms with suspense `false` that will add an `initialValue/Data` into the `atomWithObserver` that effectively disables suspense for such users.

Also, I went ahead and removed the `data` atom. Also, I removed `Action` altogether. My logic behind it is that changes are already breaking enough, so trying to keep some backward compatibility is rather makes no sense at this point. 

I also changed the API of all the atoms not to be multiple arguments, but a single object argument with multiple fields as it's easier to extend.

One thing I would wish to ship one day is `getPause` option that will allow for pausing of the query the same way as `urql` original client does. But I think it's rather a nice addition than what's absolutely needed now so I would postpone that to the next work.

RIP21 commented:
So @dai-shi I've added `getPause` functionality that is almost 1:1 to URQL bindings (slightly different as it's impossible to trigger the fetch while paused, yet possible to work around by unpausing the query, which is IMO more expected behavior) 

I also improved typings with some powerful generic types so now only needed and required types are allowed.
I also added tests and type-tests for all these things.

I think it's ready to be shipped. Please, review, comment, and merge if you can. Commit history is pretty clean, so up to you if it makes sense to squash it.

Let me know where I can contribute to the docs. Also, I'll be happy with some sort of contributor role, so I can trigger workflows and whatnot. Thanks!

RIP21 commented:
@dai-shi I need a bit of help from you.

For some reason when I make getVariables to be async function. And it makes all the atoms async as we want to await for them it in source atom to then chain all the way to the last one public to the user. But for some reason it starts to fail with `r is not a function`. Any clues?

I would want to ship this version and make it as a sepearted PR, so it's a bit clearer what I'm talking about, but if you can help in this one, then I'll be super happy to ship it with the support for async getVariables.

dai-shi commented:
> Also, I'll be happy with some sort of contributor role

Sent an invitation!

dai-shi commented on .eslintrc.json:
```diff
@@ -26,6 +26,7 @@
   "parserOptions": {
     "ecmaVersion": 2018,
     "sourceType": "module",
+    "project": "./tsconfig.json",
```
curious, if it's the default, no? I didn't know this makes any difference.

RIP21 replied:
Probably it's a default. Set it to be 100% sure. It used to be required I think.

RIP21 replied:
Probably it's a default. Set it to be 100% sure. It used to be required I think.

dai-shi commented on .github/workflows/cd.yml:
```diff
@@ -9,9 +9,19 @@ jobs:
   publish:
     runs-on: ubuntu-latest
     steps:
-      - uses: DerYeger/pnpm-setup-action@master
```
Is is just disliking someone's workflow?
Or enabling the cache? We should create our own then.

RIP21 replied:
It's because the one that is used is not allowing to select which version of PNPM to use and is 2 years without any updates. The one I used is an official Github action of PNPM

RIP21 replied:
It's because the one that is used is not allowing to select which version of PNPM to use and is 2 years without any updates. The one I used is an official Github action of PNPM

dai-shi replied:
That sounds fair. I wonder if you'd be interested in creating an updated action. Or you prefer using the official version and copying yaml configs to all jotaijs/* repos?

RIP21 replied:
I would prefer to use the official version. And about copying. I can do that, but later. I'm well overdue with this task and want to focus on work at the company where I'm working for a bit :)

dai-shi commented on .github/workflows/ci.yml:
```diff
@@ -6,9 +6,35 @@ on:
+        run: pnpm install --frozen-lockfile
+
+      - name: Install Playwright Browsers
+        run: pnpm exec playwright install --with-deps
```
Can we run playwright in a different workflow?

Our motivation is to keep same config files as much as possible along with jotaijs/*.

RIP21 replied:
We can, surely.

RIP21 replied:
Makes no sense. Look at the latest comment.

dai-shi commented on examples/01_typescript/src/App.tsx:
```diff
@@ -2,7 +2,7 @@ import React, { Suspense } from 'react'
 import { useAtom } from 'jotai/react'
 import { atom } from 'jotai/vanilla'
-import { atomsWithQuery } from 'jotai-urql'
+import { atomWithQuery } from '../../../src/index'
```
The reason we prefer `jotai-urql` is to allow running the example directly in codesandbox.

RIP21 replied:
Question then. Is it possible to upload examples from the codebase up to codesandbox and keep them in sync?

dai-shi replied:
https://githubbox.com/jotaijs/jotai-urql/tree/main/examples/01_typescript
which redirects to https://codesandbox.io/s/github/jotaijs/jotai-urql/tree/main/examples/01_typescript

RIP21 replied:
Pretty cool. Didn't know that was possible.

dai-shi commented on package.json:
```diff
@@ -28,16 +28,14 @@
-    "test": "run-s eslint tsc-test jest",
-    "eslint": "eslint --ext .js,.ts,.tsx .",
-    "jest": "jest",
+    "test": "run-s eslint tsc-test vitest playwright",
```
I see, we run playwright in "test". If we want to keep it, we can't separate workflows.

dai-shi commented on examples/01_typescript/src/App.tsx:
```diff
@@ -139,14 +139,13 @@ query($code: ID!) {
-)
+  getVariables: (get) => ({ code: get(codeAtom) }),
+  getClient: () => client,
+})
```
The reason we prefer function arguments to one object argument is to avoid property names for better minification. It's also how `jotai` is designed.
It's a subtle thing, so if you prefer an object argument, I don't mind.

RIP21 replied:
Object argument is so much better in this case due to multiple options that are optional. Surely it's possible to achieve with method overload, but that would make typings major pain to handle. So, yes, I would prefer options here to be closer to how it's done in original urql and similar to original Apollo bindings too.

dai-shi replied:
I'm fine and making is similar with urql react binding would be important.

dai-shi commented on src/atomWithMutation.ts:
```diff
@@ -0,0 +1,56 @@
+  Data = unknown,
+  Variables extends AnyVariables = AnyVariables
+>(
+  query: DocumentInput<Data, Variables>,
+  getClient: (get: Getter) => Client = (get) => get(clientAtom)
```
Hm, but these are arguments...

RIP21 replied:
Because all it takes is Document. Similar to react-urql bindings. There is no variables, there is no context, there is document and client. That's it. So making it object would be redundant.

dai-shi replied:
If this is how react-urql does, it's fine.

RIP21 replied:
Yeah it's.
https://github.com/urql-graphql/urql/blob/35657e771633be43b850f0a1f7fdbac94b075dd7/packages/react-urql/src/hooks/useMutation.ts#L146

dai-shi commented on src/atomWithMutation.ts:
```diff
@@ -0,0 +1,56 @@
+  const atomDataBase = atom<InitialOperationResult<Data, Variables>>(
+    urqlReactCompatibleInitialState
+  )
+  atomDataBase.onMount = (setAtom) => {
+    return () => {
+      setAtom(urqlReactCompatibleInitialState)
+    }
+  }
```
This looks a little weird.
onMount should be paired with onUnmount. (Like React's useEffect.)
Otherwise, this could be a source of bugs.

RIP21 replied:
It cleans up cache (only used for paused case) when user redirects from the page. Similarly to react-urql.

Unless I miss on how onMount works, this is the only way I figured how to replicate that behavior.

dai-shi replied:
What would it be if we were to setAtom on mount?

RIP21 replied:
Same thing as what it is in the cleanup callback :) We just reset it as if it was never in use.

RIP21 replied:
Renamed

dai-shi commented on src/atomWithQuery.ts:
```diff
@@ -0,0 +1,57 @@
+    getClient = (get: Getter) => get(clientAtom),
+    getPause: getPause = () => false,
+  } = options
+  const cache = new WeakMap<Client, Operation>()
```
This cache is shared across Jotai providers, unless it's your intention, we should wrap it in an atom.
```js
const cacheAtom = atom(() => new WeakMap<Client, Operation>())
```
which creates a new weak map for each provider (store).

RIP21 replied:
It gets created when atomWithQuery gets called and then it gets closured into the callbacks defined below. Every time atomWithQuery will be called, this local WeakMap will be created. I don't see how it will be shared across other atoms and providers. If that would be an atom, then it may be shared in case it's instantiated outside of the AtomWithQuery function. But in this case I don't see how it gets into the global scope of things.

dai-shi replied:
I can try creating a contrived example. This also affects the following comments.

RIP21 replied:
Removed this

dai-shi commented on src/atomWithQuery.ts:
```diff
@@ -0,0 +1,57 @@
+  // This is to avoid recreation of the client on every operation result change
+  // This is to make it more reliable when people do mistakes plus
+  // making client dynamic is rather very bad idea, as all cache is in the client
+  let client: Client
```
The same applies to this.

dai-shi replied:
Why is `client` recreated on every time?

RIP21 replied:
It is not. (aside from examples and tests)

When atomWithQuery gets called this gets closured by all the things below, again, and then gets set in the getArgs function. And then used in functions below (e.g. handleAction).

I don't understand how this variable can be used across providers when it's scoped to simple JS function. I don't see how Jotai can do some magic and collect variable declarations into some catalogue to reuse or so.

RIP21 replied:
Changed it so this client and cache weak map is not needed.

dai-shi commented on src/common.ts:
```diff
@@ -1,84 +1,74 @@
 
-    return resultAtom
+  const cacheResultAtom = atom<{ cache: Map<number, Result> }>({
+    cache: new Map(),
```
Again, this is very tricky, but `new Map()` is called during the definition, so it's shared across providers.

RIP21 replied:
Will add to onMount. Doesn't really matter. But to clarify why this is needed.

This is hack to preserve a cache of all state changes (operation results that are getting piped/streamed from source) of the query, and then show the last one, when we pause (getPause returns true) and our source becomes null (so we unsubscribe from changes).

And so instead of undefined or initial state placeholder, we show last state before the pause. In case of unpause we will subscribe to the source again, and show actual data from the urql client cache that will be piped from it through source into our observable.

RIP21 replied:
Changed that to a simple object

dai-shi commented on src/common.ts:
```diff
@@ -1,84 +1,74 @@
-        set(refreshAtom, (c) => c + 1)
-      }
-      return handleAction(action, client, refresh)
+  cacheResultAtom.onMount = (setAtom) => {
+    return () => {
+      setAtom({ cache: new Map() })
```
So, you should create a new cache on mount. Not on unmount.

RIP21 replied:
Here I'm resetting it. It's already a `new Map()` on creation. Here I just cleared away the cache when a person will redirect out of the page that uses the query. It's for `pause` functionality only.

dai-shi commented on src/common.ts:
```diff
@@ -1,84 +1,74 @@
-import { atom } from 'jotai/vanilla'
+import type { Client, OperationContext, OperationResult } from '@urql/core'
+import { AnyVariables, Operation } from '@urql/core'
+import { atomWithObservable } from 'jotai/utils'
```
```suggestion
import { atomWithObservable } from 'jotai/vanilla/utils'
```
It allows to use `jotai-urql` without React.

dai-shi commented on src/suspenseAtom.ts:
```diff
@@ -0,0 +1,3 @@
+import { atom } from 'jotai/vanilla'
+
+export const suspenseAtom = atom<boolean>(true)
```
Oh, this is a very interesting idea.

dai-shi reviewed (changes requested):
This looks very promising! So many things to learn. We should migrate to vitest in other libs and playwright would be nice too.

Found out there are some undocumented rules/conventions for developing such integration libraries.
Tried to check them in this review, but I would need to do it again.

dai-shi commented:
> For some reason when I make getVariables to be async function. And it makes all the atoms async as we want to await for them it in source atom to then chain all the way to the last one public to the user. But for some reason it starts to fail with `r is not a function`. Any clues?

Hmm, no. We need to know what is `r` by disabling minification. (we should disable minification anyways. https://github.com/dai-shi/proxy-memoize/pull/66 )

> I would want to ship this version and make it as a sepearted PR, so it's a bit clearer what I'm talking about, but if you can help in this one, then I'll be super happy to ship it with the support for async getVariables.

I don't think we could/should support async getVariables. 
In `jotai-tanstack-query`, we made a new wrapper function for async.
Issue: https://github.com/jotaijs/jotai-tanstack-query/issues/21
PR: https://github.com/jotaijs/jotai-tanstack-query/pull/30

RIP21 commented:
Thanks for comments. Will address most. Will make separate workflow etc. Will look into tanstack query thingy you mentioned.

RIP21 commented:
So, I fixed:
- jotai-urql in the example
- gave up on support of async variables, so we can ship it
- fixed ci/cd (it makes no sense to make a playwright a separate workflow as it runs the entire suite in 2 seconds on CI)
- fixed eslint stuff (removed default config and fixed alias)

dai-shi commented:
Cool. Let me go the second iteration with fresh eyes.

dai-shi commented on src/common.ts:
```diff
@@ -56,58 +71,38 @@ export const createAtoms = <
+      return result
     },
-    (_get, set, action: Action) => set(statusAtom, action)
+    (get, __, context?: Partial<OperationContext>) => {
```
Just preference.
```suggestion
    (get, _set, context?: Partial<OperationContext>) => {
```

dai-shi commented on .github/workflows/cd.yml:
```diff
@@ -9,9 +9,19 @@ jobs:
+          version: 8.2.0
+
+      - name: Use Node.js 18.x
+        uses: actions/setup-node@v3
```
I think with this, we can avoid setting `NODE_AUTH_TOKEN` by adding `registory-url` here. Let's do that. (It can be done in a follow-up PR.)

RIP21 replied:
Not sure I understand what you're talking about :)

dai-shi replied:
I'll add some commits.

dai-shi commented on .github/workflows/cd.yml:
```diff
@@ -9,9 +9,19 @@ jobs:
+        uses: actions/setup-node@v3
         with:
-          node-version: 18
+          node-version: "v18"
```
out of curiosity, how it it different?

RIP21 replied:
No difference TBH :)  Just what I've been using in my projects. Will revert to 18 :)

dai-shi reviewed (approved):
Thanks for your work. Although I don't understand 100% of changes, it looks good overall. So, let's ship it.

RIP21 commented on src/common.ts:
```diff
@@ -1,50 +1,65 @@
-import type { Client, OperationResult } from '@urql/core'
-import { atom } from 'jotai/vanilla'
+import type { Client, OperationContext, OperationResult } from '@urql/core'
+import { AnyVariables, Operation } from '@urql/core'
```
They are

RIP21 replied:
Will fix that

RIP21 commented on src/atomWithQuery.ts:
```diff
@@ -0,0 +1,51 @@
+  return createAtoms(
+    (get) => [query, getVariables(get), getContext?.(get)] as const,
+    getClient,
+    (_client, args) => {
```
Makes sense. Will rename

RIP21 commented:
@dai-shi unsure if it's a minor bump :) As APIs and other stuff got changed completely and is pretty breaking. So it's your call whether it's a minor bump or not. But for me, it sounds like a major bump :)

And yeah, if you can, please release it. As I don't have time now to figure out how it all (publishing in this project) works :)

dai-shi commented:
@RIP21 

This hasn't reached v1 yet. We add breaking changes in v0.x.

I'll merge it and make `v0.7.0` release. Just adding a tag should trigger the CD workflow.

RIP21 commented:
Gotcha. Ok then. Thanks :)
I would need to update the docs for URQL Jotai.

Also, Jotai docs seem to be outdated and are still more about v1 than v2. Any plans to update that?

dai-shi commented:
Published: https://www.npmjs.com/package/jotai-urql/v/0.7.0

(There were some errors of mine in cd.yml which is fixed in the main branch.)

sandhose commented:
Just a head's up that the published v0.7.0 has the typescript types in the wrong directory. They are in `dist/src/src/index.d.ts`, while the package.json references `dist/src/index.d.ts`

dai-shi commented:
(Ahh, we don't have tests to cover that.)

@sandhose Do you know how to fix? `dist/src/index.d.ts` should be something we want.

dai-shi commented:
Or, maybe we can fix package.json with `dist/src/src/index.d.ts` for now if it's easier.

sandhose commented:
@dai-shi I think I managed to fix it in #12
